frame/democracy/src/lib.rs

```diff
@@ -1757,12 +1754,21 @@ impl<T: Config> Pallet<T> {
 		weight = weight.saturating_add(T::WeightInfo::on_initialize_base(r));
 		// tally up votes for any expiring referenda.
 		for (index, info) in Self::maturing_referenda_at_inner(now, next..last).into_iter() {
-			let approved = Self::bake_referendum(now, index, info)?;
+			let approved = Self::bake_referendum(now, index, info);
 			ReferendumInfoOf::<T>::insert(index, ReferendumInfo::Finished { end: now, approved });
 			weight = max_block_weight;
 		}
 
-		Ok(weight)
+		<LowestUnbaked<T>>::mutate(|ref_index| {
+			while *ref_index < last &&
+				Self::referendum_info(*ref_index)
+					.map_or(true, |info| matches!(info, ReferendumInfo::Finished { .. }))
```

Comment:
I don't really like this logic.

In an on-initialize, we have some loop which can read some unbounded(?) number of items?

Reply:
note that this is what `maturing_referenda_at_inner` is doing.
`maturing_referenda_at_inner` is iterating on all the referendum from lowest unbaked to the last, and look into which one are maturing.
So this iteration to update the lowest unbaked will only iterate for a subset of the iteration in `maturing_referenda_at_inner` and so it will only trigger no more storage read but only storage read from cache.

That said I should probably refactor maturing_referenda_at_inner to give the new lowest_unbaked.
And we may want to put a bound on the maximum of ongoing referendum.

Reply:
> That said I should probably refactor maturing_referenda_at_inner to give the new lowest_unbaked. And we may want to put a bound on the maximum of ongoing referendum.

yes, or at least a bound on the max range `maturing_referenda_at_inner` will iterate over. Do we have a tracking issue for this yet?

Reply:
looking again it should already be bounded by something like `ceil(LaunchPeriod/VotingPeriod)`.

Reply:
would be nice to add a little inline comment noting that.